File: src/main/java/self/ed/generator/Generator.java
package self.ed.generator;

import static java.util.Collections.emptySet;
import static java.util.Collections.shuffle;
import static java.util.Collections.singleton;
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toSet;
import static java.util.stream.IntStream.rangeClosed;

import static self.ed.util.Utils.copy;
import static self.ed.util.Utils.countOpen;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.commons.lang3.tuple.Pair;

import self.ed.exception.ComplexityLimitException;
import self.ed.exception.MultipleSolutionsException;
import self.ed.exception.NoSolutionException;
import self.ed.solver.Cell;
import self.ed.solver.CleverSolver;

public class Generator {

  private final int size;
  private final int blockSize;
  private final Set<Integer> values;

  public Generator(int size) {
    this.size = size;
    this.blockSize = (int) Math.sqrt(size);
    this.values = rangeClosed(1, size).boxed().collect(toSet());
  }

  public Integer[][] generate() {
    return generate(size * size);
  }

  public Integer[][] generate(int complexityLowerLimit) {
    return generate(new Integer[size][size], complexityLowerLimit);
  }

  public Integer[][] reduce(Integer[][] initialValues) {
    List<Cell> open = new ArrayList<>();
    for (int row = 0; row < size; row++) {
      for (int col = 0; col < size; col++) {
        if (initialValues[row][col] != null) {
          open.add(new Cell(row, col, 0, emptySet()));
        }
      }
    }

    return reduce(initialValues, open);
  }

  private Integer[][] reduce(Integer[][] initialValues, Collection<Cell> closeCandidates) {
    Map<Cell, Integer[][]> candidates = new HashMap<>();
    closeCandidates.forEach(cell -> {
      Integer[][] nextGuess = copy(initialValues);
      nextGuess[cell.getRow()][cell.getCol()] = null;
      try {
        new CleverSolver(nextGuess).solve();
        candidates.put(cell, nextGuess);
      } catch (MultipleSolutionsException e) {
        // no-op
      }
    });

    List<Cell> nextCloseCandidates = new ArrayList<>(candidates.keySet());
    shuffle(nextCloseCandidates);
    return new ArrayList<>(nextCloseCandidates).stream()
        .map(cell -> {
          nextCloseCandidates.remove(cell);
          return reduce(candidates.get(cell), nextCloseCandidates);
        })
        .map(matrix -> Pair.of(matrix, countOpen(matrix)))
        .min(comparing(Pair::getValue))
        .map(Pair::getKey)
        .orElse(initialValues);
  }

  private Integer[][] generate(Integer[][] initialValues, int complexityLowerLimit) {
    if (countOpen(initialValues) > complexityLowerLimit) {
      throw new ComplexityLimitException();
    }

    try {
      new CleverSolver(initialValues).solve();
      return initialValues;
    } catch (MultipleSolutionsException e) {
      List<Cell> open = new ArrayList<>();
      List<Cell> pending = new ArrayList<>();
      for (int row = 0; row < size; row++) {
        for (int col = 0; col < size; col++) {
          Integer value = initialValues[row][col];
          int block = blockSize * (row / blockSize) + col / blockSize;
          if (value != null) {
            open.add(new Cell(row, col, block, singleton(value)));
          } else {
            pending.add(new Cell(row, col, block, emptySet()));
          }
        }
      }

      shuffle(pending);
      for (Cell cell : pending) {
        Integer[][] nextGuess = copy(initialValues);
        List<Integer> values = new ArrayList<>(this.values);
        open.stream().filter(cell::isRelated).map(Cell::getCandidate).forEach(values::remove);
        shuffle(values);
        for (Integer value : values) {
          nextGuess[cell.getRow()][cell.getCol()] = value;
          try {
            return generate(nextGuess, complexityLowerLimit);
          } catch (NoSolutionException e2) {
            // Our guess did not work, let's try another one
          }
        }
      }
      throw new NoSolutionException();
    }
  }
}

File: cpp-sudoku/src/generator.cpp
#include "generator.hpp"
#include "clever_solver.hpp"
#include "cell.hpp"
#include "utils.hpp"
#include <cmath>
#include <algorithm>
#include <random>

namespace sudoku {

Generator::Generator(int size) : Generator(size, size * size) {}

Generator::Generator(int size, int complexityLowerLimit) 
    : size(size), complexityLowerLimit(complexityLowerLimit), 
      blockSize(static_cast<int>(std::sqrt(size))), rng(std::random_device{}()) {
    
    for (int i = 1; i <= size; ++i) {
        values.insert(i);
    }
}

std::vector<std::vector<std::optional<int>>> Generator::generate() {
    std::vector<std::vector<std::optional<int>>> initialValues(size, std::vector<std::optional<int>>(size));
    return generate(initialValues);
}

std::vector<std::vector<std::optional<int>>> Generator::generate(const std::vector<std::vector<std::optional<int>>>& initialValues) {
    if (Utils::countOpen(initialValues) > complexityLowerLimit) {
        throw ComplexityLimitException();
    }

    try {
        CleverSolver solver(initialValues);
        solver.solve();
        // Return the original puzzle with clues, not the complete solution
        return initialValues;
    } catch (const MultipleSolutionsException& e) {
        std::vector<Cell> open;
        std::vector<Cell> pending;
        
        for (int row = 0; row < size; row++) {
            for (int col = 0; col < size; col++) {
                std::optional<int> value = initialValues[row][col];
                int block = blockSize * (row / blockSize) + col / blockSize;
                if (value.has_value()) {
                    std::unordered_set<int> candidates{value.value()};
                    open.emplace_back(row, col, block, candidates);
                } else {
                    pending.emplace_back(row, col, block);
                }
            }
        }

        // Shuffle pending cells
        std::shuffle(pending.begin(), pending.end(), rng);
        
        for (const Cell& cell : pending) {
            std::vector<std::vector<std::optional<int>>> nextGuess = Utils::copy(initialValues);
            std::vector<int> availableValues(values.begin(), values.end());
            
            // Remove values that are already used in related cells
            for (const Cell& openCell : open) {
                if (cell.isRelated(openCell)) {
                    auto candidate = openCell.getCandidate();
                    if (candidate.has_value()) {
                        auto it = std::find(availableValues.begin(), availableValues.end(), candidate.value());
                        if (it != availableValues.end()) {
                            availableValues.erase(it);
                        }
                    }
                }
            }
            
            // Shuffle available values
            std::shuffle(availableValues.begin(), availableValues.end(), rng);
            
            for (int value : availableValues) {
                nextGuess[cell.getRow()][cell.getCol()] = value;
                try {
                    return generate(nextGuess);
                } catch (const NoSolutionException& e2) {
                    // Our guess did not work, let's try another one
                }
            }
        }
        throw NoSolutionException();
    }
}

} // namespace sudoku
